### PR TITLE
WIP Traffic Ops Go ATS Configs - meta, profiles, remap.config

### PR DESCRIPTION
**Pull request by rob05c**
#### What does this PR do?

WIP Traffic Ops Go ATS Configs - meta, profiles, remap.config

WIP, just wanted to make it publicly visible. I still need to verify and do parity testing.

#### Which TC components are affected by this PR?

- [ ] Documentation
- [ ] Grove
- [ ] Traffic Analytics
- [ ] Traffic Monitor
- [x] Traffic Ops
- [ ] Traffic Ops ORT
- [ ] Traffic Portal
- [ ] Traffic Router
- [ ] Traffic Stats
- [ ] Traffic Vault
- [ ] Other _________

#### Check all that apply

- [ ] This PR includes tests
- [ ] This PR includes documentation updates
- [ ] This PR includes an update to CHANGELOG.md
- [ ] This PR includes all required license headers
- [ ] This PR includes a database migration (ensure that migration sequence is correct)
- [ ] This PR fixes a serious security flaw. Read more: [www.apache.org/security](http://www.apache.org/security/)

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2738/
Test FAILed.

**Comment by rawlinp** [comment]
I know we're still waiting on the Voluspa stuff for this, but were Go templates also considered? I figured those might be easier to maintain than all the string building done in the Perl version, but maybe there are other cons to using templates. If we did go with templates, it would probably be easier to change the format from legacy to yaml for ATS 8.

**Comment by rob05c** [comment]
I agree, and I did consider them. I'm not positive `text/template` is powerful enough, but even if it isn't, it might be worth finding a library that is.

But for now, I'd like to transliterate the Perl as closely as possible, just to minimize the risk of accidentally changing the logic. There's a lot of logic, and it's a pretty critical part of the CDN.

I'm actually hoping to do a lot of rewriting, once it's in Go, and easier to understand and work with. I'm also really hoping, after it's all in Go, to go back and analyze "these are all the queries, and this is all that data needed by a config file / cache", and then make a new TO endpoint that contains all that data, as a generic "cache config data" endpoint.

That lets us do 2 things. We can then write a library that hits that endpoint over HTTP and builds the config files from it, which can be used by ORT. But as an intermediary step, we can also change the config file endpoints to use that single "getConfigFileData" func and struct, and then build the config files to be served using that data as well. In fact, we can do both, reducing duplicate code, as well as the total amount of SQL code, and likely also the number of queries, improving performance.

Which, I'm also hoping that single "cache config data" endpoint can be used by Voluspa, and also serve as the "generic CDN cache config language/protocol" we've been wanting to make for a long time now. (Though I think the common practice of calling it a "language" or "protocol" are misusing the words, notwithstanding).

And I'm guessing the best way to build those config files from the "cache data" endpoint, is going to be with templates.

So yeah, hoping to do all those things. Just hoping to do them in increments, so each step is smaller and safer.

**Comment by rawlinp** [comment]
One thing I'd like to consider for ORT 2.0 is versioning each cache's config using git via TO's "queue updates" API and having ORT 2.0 do a `git pull` to just get file diffs rather than entire files every time (or all the required data to generate its own config every time). That would bring us some benefits but also some interesting challenges as well.

**Comment by ocket8888** [comment]
Regarding that, I had a similar idea - if profiles and parameters defining a config haven't been modified, TO could return a `304 NOT MODIFIED` in response to a request using a `IF-MODIFIED-SINCE` header. It would mean that when a file is updated you pull the entire thing, not just diffs, but would allow TO to skip generation in cases where it hasn't been modified.

Also, the way our config file generation works atm (may change when re-written to Go) is non-deterministic, and can result in diffs that aren't real - just different ordering of the same lines.

**Comment by rawlinp** [comment]
Yeah, if we're keeping it an HTTP API we should try to implement proper cache-control if we can. The hard thing there is that one little change in the API can invalidate a lot of config across the CDN, and you're typically only queuing updates on caches that need to pull a new version of their config. And we need a good way to track versions of configs across all the caches in the CDN, which is where I think Git would come in. If we didn't persist the different versions of the configs in TO, we'd probably still have to generate the file on request, take the hash, compare that to the request's hash, and 304 if it's the same or 200 if it's different. At that point we're only saving bandwidth on the network as opposed to load on TO/TODB as well since we'd have to generate the file for every request anyways. But Git would bring in its own set of challenges like figuring out which files actually need updated when the API is updated.

Anyways, if we could boil all the required data for generating a cache's config down into a single endpoint like Rob said, we would probably get a lot of bang for our buck and would probably be able to just add a timestamp to the resulting data from that endpoint so that one "Queue Update" command only requires digging that data up one time, allowing all the caches affected by that "Queue Update" to just grab that single version of the data which could be easily cached.

As I type this out, it's sounding more and more like CRConfig snapshots but for cache config. Maybe that wouldn't be such a bad thing.

**Comment by rob05c** [comment]
>if profiles and parameters defining a config haven't been modified, TO could return a 304 NOT MODIFIED
I agree, it would be great for TO to return 304 Not Modified for all endpoints, especially big things like the CRConfig.

But in this particular case, I don't think you'd really save much computation (just network). The computation isn't the majority of the time spent, the majority of time is querying the data. Either way, IMO we should transliterate the Perl first, then look at ways to improve performance.

**Comment by rob05c** [comment]
>As I type this out, it's sounding more and more like CRConfig snapshots but for cache config. Maybe that wouldn't be such a bad thing.

Server Snapshots is already a necessary part of the Self-Service plan. See https://cwiki.apache.org/confluence/display/TC/Traffic+Control++Self+Service+Proposal+for+Change+Integrity

That's one of the powerful things about using Timestamps to solve the SS config deployment problem: it makes things like Cache-Control, 304, and diffs (HTTP Delta / JSON Patch) incredibly easy.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2846/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3276/
Test FAILed.

File: lib/go-tc/ats.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package tc
```
Comment (ocket8888) [inline lib/go-tc/ats.go]: Missing license header - this is why the Jenkins job failed.

**Comment by ocket8888** [comment]
Looks like the meta path has dropped the ability to parse server names as opposed to only integral, unique identifiers:

```console
$ export R_PATH="api/1.3/servers/$SERVER_NAME/configfiles/ats" # Server name omitted
$ export TO_URL="https://localhost:6443" # URL to the current TO that uses Perl for configs
$ toget -fpr "$R_PATH" # Credentials omitted for obvious reasons
GET /api/1.3/servers/REDACTED/configfiles/ats HTTP/1.1
Host: trafficops.infra.ciab.test
User-Agent: curl/7.47.0
Accept: */*
Cookie: mojolicious=...

HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept
Access-Control-Allow-Methods: POST,GET,OPTIONS,PUT,DELETE
Access-Control-Allow-Origin: *
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: text/plain;charset=UTF-8
Date: Tue, 05 Mar 2019 19:31:55 GMT
Server: Mojolicious (Perl)
Set-Cookie: mojolicious=...; expires=Tue, 05 Mar 2019 23:31:55 GMT; path=/; HttpOnly
Vary: Accept-Encoding
Whole-Content-Sha512: sMhpmferOoh/fcxDBG7ZT8f5Zt++0b9j7iqrH8coJ15E1Qyf2yrqXoCwJ9KOkTfYrGxrsKy0wMhuoRsDllOMgQ==
Transfer-Encoding: chunked

{
	"info": {
		"profileId": 390,
		"toRevProxyUrl": "REDACTED",
		"toUrl": "REDACTED",
		"serverIpv4": "REDACTED",
		"serverTcpPort": 80,
		"serverName": "REDACTED",
		"cdnId": 1,
		"cdnName": "REDACTED",
		"serverId": 873,
		"profileName": "REDACTED"
	},
	"configFiles": [
		{
			"fnameOnDisk": "12M_facts",
			"location": "/opt/ort",
			"apiUri": "/api/1.2/profiles/REDACTED/configfiles/ats/12M_facts",
			"scope": "profiles"
		},
		"scores of other files omitted for length"
	]
}
$ export TO_URL="https://localhost:6444" # URL to a TO build using this PR
$ toget -fpr "$R_PATH"
GET /api/1.3/servers/REDACTED/configfiles/ats HTTP/1.1
Host: trafficops.infra.ciab.test
User-Agent: curl/7.47.0
Accept: */*
Cookie: mojolicious=...

HTTP/1.1 400 Bad Request
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Set-Cookie, Cookie
Access-Control-Allow-Methods: POST,GET,OPTIONS,PUT,DELETE
Access-Control-Allow-Origin: *
Content-Type: application/json
Set-Cookie: mojolicious=...; Path=/; HttpOnly
Whole-Content-Sha512: 0RV1+AA8FYWqzq40fM/BnkLYpBUzloc2JOeuxEu/m8Wjqoha1kKJwXEtBcMicN6VVQHsSGQIWBwj2HdVVjcOCQ==
X-Server-Name: traffic_ops_golang/
Date: Tue, 05 Mar 2019 19:42:35 GMT
Content-Length: 97

{
	"alerts": [
		{
			"text": "getting integer parameters: parameter 'id' not an integer",
			"level": "error"
		}
	]
}
```

File: lib/go-tc/ats.go (hunk 1 above)
Comment (rob05c) [reply lib/go-tc/ats.go]: Fixed

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3567/
Test FAILed.

**Comment by rob05c** [comment]
This has been separated into:
https://github.com/apache/trafficcontrol/pull/3516
https://github.com/apache/trafficcontrol/pull/3593
https://github.com/apache/trafficcontrol/pull/3688

**Pull request closed without merging**